### PR TITLE
JB:10 Character creator - fixes codereview

### PR DESCRIPTION
**Pull request by Jakubba**

### CODE REVIEW

File: src/components/Equipment.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    <div>
+      <h2>Ekwipunek</h2>
+      <ItemList items={availableItems} onAddItem={handleAddItem} />
+      <div>
+        <h3>Twoje ekwipunek</h3>
+        <ul>
+          {items.map((item, index) => (
+            <li key={index}>
+              <img src={item.image} alt={item.name} className="w-16 h-16" />
+              <p>{item.name}</p>
```
Comment (fengerkamil) [inline src/components/Equipment.tsx]: Tutaj mogą być lepsze tagi `<figure>` `` oraz `<figcaption>`

File: src/components/Equipment.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import React, { useState } from 'react';
+import ItemList from './ItemList';
+import torches from './../assets/image/ekwipunek/torches.png';
+
+const Equipment = () => {
+  const [items, setItems] = useState([]);
```
Comment (fengerkamil) [inline src/components/Equipment.tsx]: Tutaj typy normalnie działają? Jak iterujesz po itemsach to podpowiada Ci VSC property?

File: src/components/ItemList.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import React, { useState } from 'react';
+const ItemList = ({ isOpen, onClose, items, onAddItem }) => {
+  if (!isOpen) return null;
+
+  return (
+    <div className="fixed inset-0 flex items-center justify-center bg-black bg-opacity-50">
```
Comment (fengerkamil) [inline src/components/ItemList.tsx]: To może być prawdopodobnie znacznik `<section>`

File: src/components/ItemList.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      <div className="relative w-3/4 p-4 bg-white rounded shadow-lg">
+        <button
+          onClick={onClose}
+          className="absolute p-2 text-white bg-red-500 rounded-full top-2 right-2"
+        >
+          X
+        </button>
```
Comment (fengerkamil) [inline src/components/ItemList.tsx]: Do buttona dodaj proszę atrybut `aria-label` który będzie przydatny dla czytników ekranowych (accessibility) na przyklad "Close button" + dodaj span który opakuje "X" i dodaj mu aria-hidden="true" aby ukryć tą literkę z accessibility tree, wtedy czytniki ekranowe nie odcztają userowi słowa "X" tylko to co będzie w aria-label

File: src/components/ItemList.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                className="object-contain w-full h-full mb-2"
+              />
+              <div className="absolute bottom-0 left-0 right-0 p-2 bg-white bg-opacity-75">
+                <p className="text-sm text-gray-700">
+                  Siła: {item.stats.strength}
+                </p>
+                <p className="text-sm text-gray-700">Moc: {item.stats.power}</p>
+                <button
+                  onClick={() => onAddItem(item)}
+                  className="py-1 mt-2 text-white bg-green-500 rounded-full"
```
Comment (fengerkamil) [inline src/components/ItemList.tsx]: Tutaj również możesz dodać etykietę aria-label która będzie bardziej słowna czyli "Dodaj element" albo coś.

File: src/components/Preloader.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        @keyframes spin {
+          0% {
+            transform: rotate(0deg);
+          }
+          100% {
+            transform: rotate(360deg);
+          }
+        }
+      `}</style>
+    </div>
```
Comment (fengerkamil) [inline src/components/Preloader.tsx]: To już chyba łatwiej wynieść od jakiegoś globalnego pliku css niż kisić to w JSX.

File: src/components/Weapons.tsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      image: '/src/assets/image/wapons/redsmith-hammer.png',
+      stats: { strength: 4, power: 3 },
+    },
+    {
+      id: 10,
+      name: 'Vorpal Sword',
+      image: '/src/assets/image/wapons/vorpal-sword.jpeg',
+      stats: { strength: 5, power: 5 },
+    },
+  ];
```
Comment (fengerkamil) [inline src/components/Weapons.tsx]: Takie domyślne obiekty lub arrajki przeniósłbym do globalnego folderu constants, tak samo w poprzednim pliku gdzie miałeś też jakiś jeden availableItem.

File: src/components/Weapons.tsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    const loadItems = async () => {
+      const savedItems = await getItemsFromFirestore();
+      setItems(savedItems);
+    };
+
+    loadItems();
+  }, []);
+
+  return (
+    <div>
```
Comment (fengerkamil) [inline src/components/Weapons.tsx]: `<section>`

File: src/components/Weapons.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        Dodaj przedmiot
+      </button>
+
+      <div>
+        <h3>Twoja broń i tarcze</h3>
+        <ul>
+          {items.length === 0 ? (
+            <p>Nie dodałeś jeszcze żadnych przedmiotów.</p>
+          ) : (
+            items.map((item, index) => (
```
Comment (fengerkamil) [inline src/components/Weapons.tsx]: Na tym poziomie w sumie możemy już zdestrukturyzować image, name oraz stats, aby nie powielac `item` w returnie map

File: src/components/Wizards.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      image: wizard1,
+      stats: { strength: 0, power: 5 },
+    },
+    {
+      id: 2,
+      name: 'Ice Shield',
+      image: wizard2,
+      stats: { strength: 2, power: 2 },
+    },
+  ];
```
Comment (fengerkamil) [inline src/components/Wizards.tsx]: Do osobnego pliku najlepiej, folder constants

File: src/components/Wizards.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  };
+
+  return (
+    <div>
+      <h2>Zaklęcia</h2>
+      <ItemList items={availableItems} onAddItem={handleAddItem} />
+      <div>
+        <h3>Twoje zaklęcia</h3>
+        <ul>
+          {items.map((item, index) => (
```
Comment (fengerkamil) [inline src/components/Wizards.tsx]: Destrukturyzacja

File: src/helpers/errorMessages.tsx [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    case 'auth/email-already-in-use':
+      return 'Ten e-mail jest już używany.';
+    case 'auth/invalid-email':
+      return 'Niepoprawny adres e-mail.';
+    case 'auth/weak-password':
+      return 'Hasło jest zbyt słabe.';
+    default:
+      return 'Wystąpił nieznany błąd. Spróbuj ponownie później.';
+  }
+};
```
Comment (fengerkamil) [inline src/helpers/errorMessages.tsx:12]: To jest spoko

File: src/pages/Character.tsx [hunk 13]
```diff
@@ -3,17 +3,18 @@ import { useNavigate } from 'react-router-dom';
...
 import minus from './../assets/icons/icon-minus.png';
-import weapon from './../assets/image/wapons/pngegg.png';
-import shield from './../assets/image/wapons/shield.png';
 import wizard1 from './../assets/image/zaklecia/wizards1.png';
 import wizard2 from './../assets/image/zaklecia/wiazard2.png';
 import torches from './../assets/image/ekwipunek/torches.png';
 
 const Character = () => {
   const { logout, currentUser } = useAuth();
+  const [showPop, setShowPop] = useState(false);
```
Comment (fengerkamil) [inline src/pages/Character.tsx]: A co to znaczy showPop? unikajmy skrótowców bo ciężko je zrozumieć.

File: src/pages/Character.tsx [hunk 14]
```diff
@@ -87,7 +89,14 @@ const Character = () => {
...
-
+  const handleImageUpload = (event) => {
+    const file = event.target.files[0];
+    if (file) {
+      const reader = new FileReader();
+      reader.onloadend = () => setImage(reader.result);
+      reader.readAsDataURL(file);
+    }
+  };
   const tabs = [
```
Comment (fengerkamil) [inline src/pages/Character.tsx]: To też może być w osobnym pliku w folderze constants

File: src/pages/Character.tsx [hunk 15]
```diff
@@ -124,11 +133,27 @@ const Character = () => {
             Witaj, {currentUser?.email}!
```
Comment (fengerkamil) [inline src/pages/Character.tsx]: A co jeśli nie będzie emaila? Mamy tutaj "?" więc to możliwe, musimy coś alternatywnego ustawić na ten przypadek.

File: src/pages/Character.tsx [hunk 16]
```diff
@@ -124,11 +133,27 @@ const Character = () => {
...
+                  style={{
+                    position: 'absolute',
+                    objectFit: 'contain',
+                    height: '100%',
+                    width: '100%',
+                    top: '50%',
+                    left: '50%',
+                    transform: 'translate(-50%, -50%)',
+                    zIndex: 100,
+                  }}
```
Comment (fengerkamil) [inline src/pages/Character.tsx]: Są jakieś przeciwwskazania aby użyć tutaj Tailwinda?

Comment (fengerkamil) [reply src/pages/Character.tsx]: Poniżej masz input w którym udało się to osiągnąć Tailwindem :D

File: src/pages/LoginForm/LoginForm.tsx [hunk 17]
```diff
@@ -21,17 +20,22 @@ const initialValues = {
...
+
   useEffect(() => {
     const unsubscribe = onAuthStateChanged(auth, (user) => {
       if (user) {
         navigate('/character');
+      } else {
+        setLoading(false);
       }
     });
     return () => unsubscribe();
```
Comment (fengerkamil) [inline src/pages/LoginForm/LoginForm.tsx]: `return unsubscribe` powinno wystarczyć

File: src/pages/LoginForm/LoginForm.tsx [hunk 18]
```diff
@@ -42,75 +46,86 @@ const LoginForm = () => {
...
-                    </Link>
-                  </div>
-                </Form>
-              )}
-            </Formik>
-          </section>
-        </main>
-        <InfoPage />
+        {loading && <Preloader />}
+        {!loading && (
```
Comment (fengerkamil) [inline src/pages/LoginForm/LoginForm.tsx]: To możemy połączyć `{ loading ? <Preloader /> : <main> ... </main> }`

File: src/pages/LoginForm/LoginForm.tsx [hunk 19]
```diff
@@ -42,75 +46,86 @@ const LoginForm = () => {
...
+          <>
+            <main className="flex w-full min-h-[560px] h-[80vh] relative">
+              <img
+                className="absolute object-cover w-full h-full top-1/2 left-1/2 translate-x-[-50%] translate-y-[-50%]"
+                src={bgLogin}
+                alt="Background"
+              />
```
Comment (fengerkamil) [inline src/pages/LoginForm/LoginForm.tsx]: To moglibyśmy wydzielić do osobnego komponentu dziecka dla LoginForm, np LoginFormBackground

File: src/pages/LoginForm/LoginForm.tsx [hunk 20]
```diff
@@ -42,75 +46,86 @@ const LoginForm = () => {
...
+                          className="font-semibold font-secondaryFont text-primaryColor"
+                          to="/registration"
+                        >
+                          {' '}
+                          Zarejestruj się
+                        </Link>
+                      </div>
+                    </Form>
+                  )}
+                </Formik>
```
Comment (fengerkamil) [inline src/pages/LoginForm/LoginForm.tsx]: Formularz również mógłby byc osobnym komponentem LoginForm (A ten w którym jesteśmy zmieniłby nazwę np. na LoginPage). Chodzi o wydzielenie logiki formularza gdzieś obok.

File: src/pages/Registration/Registration.tsx [hunk 21]
```diff
@@ -23,16 +21,26 @@ const initialValues = {
 
 const Registration = () => {
```
Comment (fengerkamil) [inline src/pages/Registration/Registration.tsx]: RegisterPage nazwałbym, może brzmieć lepiej

File: src/pages/Registration/Registration.tsx [hunk 22]
```diff
@@ -54,11 +62,11 @@ const Registration = () => {
               validationSchema={validationSchema}
               onSubmit={handleSubmit}
             >
-              {() => (
+              {({ isSubmitting }) => (
```
Comment (fengerkamil) [inline src/pages/Registration/Registration.tsx]: Tutaj również formularz do przeniesienia RegisterForm komponent dedykowany, będzie bardziej modularnie, mniejsze komponenty

File: src/pages/Registration/Registration.tsx [hunk 23]
```diff
@@ -93,8 +101,9 @@ const Registration = () => {
                     type="submit"
                     className="px-4 py-2 mt-4 font-semibold text-white rounded-lg font-secondaryFont bg-thirdColor hover:bg-fourthColor focus:outline-none focus:ring-2 focus:bg-fourthColor"
                     aria-label="Register Account"
+                    disabled={loading || isSubmitting}
                   >
-                    Załóż konto
+                    {loading ? 'Ładowanie...' : 'Zarejestruj się'}
```
Comment (fengerkamil) [inline src/pages/Registration/Registration.tsx]: Tutaj jeszcze przydałby sie dla tego buttona `aria-live`, na zadanie domowe zerknij co ono wnosi:)

File: src/pages/Registration/Registration.tsx [hunk 24]
```diff
@@ -93,8 +101,9 @@ const Registration = () => {
                     type="submit"
                     className="px-4 py-2 mt-4 font-semibold text-white rounded-lg font-secondaryFont bg-thirdColor hover:bg-fourthColor focus:outline-none focus:ring-2 focus:bg-fourthColor"
                     aria-label="Register Account"
```
Comment (fengerkamil) [inline src/pages/Registration/Registration.tsx]: On jest raczej będny bo tekst buttona jest sensowny

File: src/pages/LoginForm/LoginForm.tsx [hunk 25]
```diff
@@ -42,75 +46,86 @@ const LoginForm = () => {
...
       console.error('Logowanie nie powiodło się', error);
       alert('Logowanie nie powiodło się. Sprawdź swoje dane.');
+    } finally {
+      setLoading(false);
+      setSubmitting(false);
     }
   };
 
   return (
     <RedirectIfAuthenticated>
```
Comment (fengerkamil) [inline src/pages/LoginForm/LoginForm.tsx]: Może nazwać by ten komponent bardziej AuthenticationGuard?

File: src/components/Wizards.tsx [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+    setItems((prevItems) => [...prevItems, item]);
+  };
+
+  return (
+    <div>
+      <h2>Zaklęcia</h2>
+      <ItemList items={availableItems} onAddItem={handleAddItem} />
+      <div>
+        <h3>Twoje zaklęcia</h3>
+        <ul>
```
Comment (fengerkamil) [inline src/components/Wizards.tsx]: Ogólnie mam wrażenie że te komponenty są bardzo podobne. Wizards, Equipment, Weapons... moglibysmy stworzyć jeden który przyjmie odpowiednie propsy (title, subtitle, initialItems, oraz opcjonalne elementy JSX ktore beda wyrenderowane w miejscu (tam gdzie Dodaj przedmiot oraz na dole gdziej est <ItemList>

File: src/components/Wizards.tsx [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+      id: 2,
+      name: 'Ice Shield',
+      image: wizard2,
+      stats: { strength: 2, power: 2 },
+    },
+  ];
+
+  const handleAddItem = (item) => {
+    setItems((prevItems) => [...prevItems, item]);
+  };
```
Comment (fengerkamil) [inline src/components/Wizards.tsx]: Nie sprawdzamy czy już podana nazwa nie istnieje, lub to samo id? Nie chcesz żadnych warunków ustalać?

File: src/assets/image/ekwipunek/equipment1.png
Comment (fengerkamil) [inline src/assets/image/ekwipunek/equipment1.png:1]: Każde zdjęcie z ekwipunku (zmień nazwę folderu na polski) jest w dość dużym rozmiarze, czy będziesz je wyświetlał aż takie? Pewnie to będzie użyte jako mały obrazek elementów ekwipunku, więc czy jest sens ładować aż tak duże (roździelczość) obrazki? Przy okazji, im mniejszy obrazek, tym mniej waży. 

User musi mieć możliwość szybkiego załądowania strony, jak każdy element ekwipunku będzie miał 0,5MB to szybko user straci transfer :D

File: src/components/ItemsSection.tsx [hunk 28]
```diff
@@ -48,9 +59,9 @@ const ItemsSection = ({ title, itemsData }) => {
           {items?.length === 0 ? (
             <p>Nie dodałeś jeszcze żadnych przedmiotów.</p>
           ) : (
-            items.map(({ image, name, stats }, index) => (
+            items.map(({ image, name, stats, id }, index) => (
               <li
-                key={index}
+                key={id || index} // Use 'id' if available
```
Comment (fengerkamil) [inline src/components/ItemsSection.tsx]: Istnieje szansa że nie będzie ID?

File: src/firebase-function.tsx [hunk 29]
```diff
@@ -1,21 +1,94 @@
 import { collection, addDoc, getDocs } from 'firebase/firestore';
```
Comment (fengerkamil) [inline src/firebase-function.tsx]: Zbyt wiele funkcji w jednym pliku, przygotowałbym osobny folder src/services/nazwa-funkcji.ts

Comment (fengerkamil) [reply src/firebase-function.tsx]: I każde wrzucił osobno

File: src/firebase.js [hunk 30]
```diff
@@ -15,13 +13,9 @@ const firebaseConfig = {
   measurementId: 'G-80PYR1XPLM',
 };
```
Comment (fengerkamil) [inline src/firebase.js:14]: Config musimy pilnie przenieść do plików .env, nie powinny one nigdy trafić do repozytorium.
Plik .env powinien być dodany do .gitignore! Pisz priv jeśli nie wiesz jak to ogarnąć
